pkg/soc/soc.go

```diff
@@ -0,0 +1,243 @@
+// license that can be found in the LICENSE file.
+
+// Package swarm contains most basic and general Swarm concepts.
+package soc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/swarm"
+	bmtlegacy "github.com/ethersphere/bmt/legacy"
+)
+
+const (
+	IdSize        = 32
+	SignatureSize = 65
+	AddressSize   = crypto.AddressSize
+	minUpdateSize = IdSize + SignatureSize + swarm.SpanSize
+)
+
+// Id is a soc identifier
+type Id []byte
+
+// Owner is a wrapper that enforces valid length address of soc owner.
+type Owner struct {
+	address []byte
+}
+
+// NewOwner creates a new Owner.
+func NewOwner(address []byte) (*Owner, error) {
+	if len(address) != AddressSize {
+		return nil, fmt.Errorf("invalid address %x", address)
+	}
+	return &Owner{
+		address: address,
+	}, nil
+}
+
+// Update wraps a single soc update.
+type Update struct {
+	id        Id
+	span      int64
+	payload   []byte
+	signature []byte
+	signer    crypto.Signer
+	owner     *Owner
+}
+
+// NewUpdate creates a new Update from arbitrary soc id and
+// soc byte payload.
+//
+// By default the span of the soc data is set to the length
+// of the payload.
+func NewUpdate(id Id, payload []byte) *Update {
+	return &Update{
+		id:      id,
+		payload: payload,
+		span:    int64(len(payload)),
+	}
+}
+
+// WithSpan is a chainable function that adds an arbitrary
+// span to the chunk data defined for the update.
+func (s *Update) WithSpan(span int64) *Update {
+	s.span = span
+	return s
+}
+
+// WithOwnerAddress provides the possibility of setting the ethereum
+// address for the owner of an update in the absence of a signer.
+func (s *Update) WithOwnerAddress(ownerAddress *Owner) *Update {
+	s.owner = ownerAddress
+	return s
+}
+
+// AddSigner currently sets a single signer for the soc update.
+//
+// This method will overwrite any value set with WithOwnerAddress with
+// the address derived from the given signer.
+func (s *Update) AddSigner(signer crypto.Signer) error {
+	publicKey, err := signer.PublicKey()
+	if err != nil {
+		return err
+	}
+	ownerAddressBytes, err := crypto.NewEthereumAddress(*publicKey)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(os.Stderr, "%x", ownerAddressBytes)
+	ownerAddress, err := NewOwner(ownerAddressBytes)
+	if err != nil {
+		return err
+	}
+	s.signer = signer
+	s.owner = ownerAddress
+	return nil
+}
+
+// OwnerAddress returns the ethereum address of the signer of the Update.
+func (s *Update) OwnerAddress() []byte {
+	return s.owner.address
+}
+
+// Address returns the soc Chunk address of the update.
+func (s *Update) Address() (swarm.Address, error) {
+	var err error
+	hasher := swarm.NewHasher()
+	_, err = hasher.Write(s.id)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	_, err = hasher.Write(s.owner.address)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	addressBytes := hasher.Sum(nil)
+	return swarm.NewAddress(addressBytes), nil
```

Comment:
```suggestion
	return swarm.NewAddress(hasher.Sum(nil)), nil
```

Reply:
terribly sorry but I find it give better clarity to name each step of mutations as much as possible in any code.